DESIGN.md

```diff
@@ -0,0 +1,73 @@
+
+This problem is exponentially grows when we would like to run a 3rd Party plugin from the community.  
+As OpenSearch and plugins run in the same process, it brings in security risk, dependency conflicts and reduces the velocity of releases.
+
+Introducing extensions, a simple and easy way to extend features of OpenSearch. It would support all plugin features and enable them to run in a different process or another node via OpenSearch SDK.
+
+Meta Issue: [Steps to make OpenSearch extensible](https://github.com/opensearch-project/OpenSearch/issues/2447)  
+Sandboxing: [Step towards modular architecture in OpenSearch](https://github.com/opensearch-project/OpenSearch/issues/1422)  
+Security: [Security for extensions](SECURITY.md) 
+
+## Plugins Architecture
+
+![](Docs/plugins.png)
+
+Plugins are installed via [`opensearch-plugin`](https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java) and are class loaded into OpenSearch.
+Plugins run within OpenSearch as a single process. Plugins interface with OpenSearch via extension points which plug into the core modules of OpenSearch.
+This [blog post](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/) helps untangle how plugins work. 
+
+Walking through an example, a Plugin would like to register a custom setting which could be toggled via Rest API by the user. 
+The plugin uses compiles with OpenSearch x.y.z version and generates a `.zip`.   
+This `.zip` file is installed via `opensearch-plugin` tool which unpacks the code and places it under `~/plugins/<plugin-name>`.
+During the bootstrap of OpenSearch node, it class loads all the code under `~/plugins/` directory. `Node.java` makes a call to get all settings the plugins would like to register. These settings are used as `additionalSettings` and construct `SettingsModule` instance which tracks all settings.
+
+## Extensions Architecture
+
+![](Docs/Extensions.png)
+
+Extensions are independent processes which are built using `opensearch-sdk`. They communicate with OpenSearch via [transport](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-netty4) protocol which today is used to communicate between OpenSearch nodes.  
+Extensions are designed to extend features via APIs which are exposed using extension points of OpenSearch.  
+
+### Discovery
+Extensions are discovered and configured via `extensions.yml` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.
+Here is an example extension configuration `extensions.yml`:
+
+```
+extensions:
+  - name: opensearch-sdk
+    uniqueId: opensearch-sdk-1
+    hostName: 'sdk_host'
+    hostAddress: '127.0.0.1'
+    port: '4532'
+    version: '1.0'
+    description: Extension for the Opensearch SDK Repo
+    opensearchVersion: '3.0.0'
```

Comment:
Why does the extension need to provide an opensearch version in this configuration?  Shouldn't this be part of the handshake during the discovery/initialization process?

Reply:
I think this is because in future a given extension can work with several versions of OpenSearch and not tied to one, this is around API versioning. Hence providing input on what version of OpenSearch a given extension can work would need to be specified.

Reply:
Thats a good point. For now its static, and configured via `yml`. But its definitely possible to transmit this information over the wire during initialization. Its too early for extensions but we can take this as a feature request.

Reply:
Can we remove this until we need it or know how it should be defined? I understand this is the design documentation whereas there has been work done to implementation, but it should hold up to scrutiny we can always create issues to clean up the implementation.

Reply:
Sure.

Reply:
Opened up an issue: https://github.com/opensearch-project/opensearch-sdk/issues/65